src/DynamicData.Tests/Cache/ExpireAfterFixture.ForSource.cs

```diff
@@ -644,77 +649,121 @@ public void TimeSelectorThrows_ErrorIsPropagated()
+            {
+                source.Edit(updater =>
                 {
                     var changeCount = randomizer.Int(1, maxChangeCount);
-
                     for (var i = 0; i < changeCount; ++i)
                     {
-                        var changeReason = (mutator.Count is 0)
-                            ? ChangeReason.Add
-                            : randomizer.Enum(exclude: ChangeReason.Moved);
-
-                        if (changeReason is ChangeReason.Add)
+                        var changeReason = randomizer.WeightedRandom(changeReasons, updater.Count switch
                         {
-                            mutator.AddOrUpdate(new Item()
-                            {
-                                Id = nextItemId++,
-                                Expiration = GenerateExpiration()
-                            });
-                            continue;
-                        }
-
-                        var key = randomizer.CollectionItem((ICollection<int>)mutator.Keys);
+                            0   => changeReasonWeightsWhenCountIs0,
+                            _   => changeReasonWeightsOtherwise
+                        });
 
                         switch (changeReason)
                         {
+                            case ChangeReason.Add:
+                                updater.AddOrUpdate(items[nextItemIndex++]);
+                                break;
+
                             case ChangeReason.Refresh:
-                                mutator.Refresh(key);
+                                updater.Refresh(updater.Keys.ElementAt(randomizer.Int(0, updater.Count - 1)));
                                 break;
 
                             case ChangeReason.Remove:
-                                mutator.RemoveKey(key);
+                                updater.RemoveKey(updater.Keys.ElementAt(randomizer.Int(0, updater.Count - 1)));
                                 break;
 
                             case ChangeReason.Update:
-                                source.AddOrUpdate(new Item()
+                                updater.AddOrUpdate(new StressItem()
                                 {
-                                    Id = key,
-                                    Expiration = GenerateExpiration()
+                                    Id          = updater.Keys.ElementAt(randomizer.Int(0, updater.Count - 1)),
+                                    Lifetime    = randomizer.Bool()
+                                        ? TimeSpan.FromTicks(randomizer.Long(minItemLifetime.Ticks, maxItemLifetime.Ticks))
+                                        : null
                                 });
                                 break;
                         }
                     }
                 });
+            }
+        }
 
-            return maxExpiration;
-
-            DateTimeOffset? GenerateExpiration()
+        private static async Task WaitForCompletionAsync(
```

Comment:
I've come across this issue in my day job recently and this looks like a good solution.

Reply:
I'd still rather have a system that synchronizes DIRECTLY, I.E. having the test somehow track every item that needs to be expired, and then manually watching for all of them to expire, but when I sat down to it, I couldn't justify it being worth all the extra code, over this simple little method.